pkg/tempo/servicegraphprocessor/processor.go

```diff
@@ -0,0 +1,256 @@
+		Help: "Total count of spans processed that were not tagged with span.kind",
+	}, []string{"span_kind"})
+
+	cs := []prometheus.Collector{
+		p.serviceGraphRequestTotal,
+		p.serviceGraphRequestFailedTotal,
+		p.serviceGraphRequestHistogram,
+		p.serviceGraphUnpairedSpansTotal,
+		p.serviceGraphUntaggedSpansTotal,
+	}
+
+	for _, c := range cs {
+		if err := p.reg.Register(c); err != nil {
+			return err
+		}
+	}
+
+	// Collect unpaired spans when evicting items from the store during
+	// periodic cleanup
+	p.store.OnEvicted(func(s string, i interface{}) {
+		e := i.(edge)
+		if !e.complete() {
+			p.serviceGraphUnpairedSpansTotal.WithLabelValues(e.clientService, e.serverService).Inc()
+		}
+	})
+
+	return nil
+}
+
+func (p *processor) Shutdown(context.Context) error {
+	p.unregisterMetrics()
+	return nil
+}
+
+func (p *processor) unregisterMetrics() {
+	cs := []prometheus.Collector{
+		p.serviceGraphRequestTotal,
+		p.serviceGraphRequestFailedTotal,
+		p.serviceGraphRequestHistogram,
+		p.serviceGraphUnpairedSpansTotal,
+		p.serviceGraphUntaggedSpansTotal,
+	}
+
+	for _, c := range cs {
+		p.reg.Unregister(c)
+	}
+}
+
+func (p *processor) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{}
+}
+
+func (p *processor) ConsumeTraces(ctx context.Context, td pdata.Traces) error {
+	level.Debug(p.logger).Log("msg", "consuming traces")
+
+	var errs error
+	for _, trace := range batchpersignal.SplitTraces(td) {
+		if err := p.consume(trace); err != nil {
+			if errors.Is(err, ErrTooManyItems) {
+				level.Warn(p.logger).Log("msg", "skipped processing of spans", "maxItems", p.maxItems, "err", ErrTooManyItems)
+				break
+			}
+			errs = multierror.Append(errs, err)
+		}
+	}
+	if errs != nil {
+		level.Error(p.logger).Log("msg", "failed consuming traces", "err", errs)
+	}
+
+	p.collectMetrics()
+
+	return p.nextConsumer.ConsumeTraces(ctx, td)
+}
+
+func (p *processor) collectMetrics() {
+	for k, v := range p.store.Items() {
+		e := v.Object.(edge)
+		if e.complete() {
+			p.serviceGraphRequestTotal.WithLabelValues(e.clientService, e.serverService).Inc()
+			if e.failed {
+				p.serviceGraphRequestFailedTotal.WithLabelValues(e.clientService, e.serverService).Inc()
+			}
+			p.serviceGraphRequestHistogram.WithLabelValues(e.clientService, e.serverService).Observe(e.serverLatency.Seconds())
+			p.store.Delete(k)
+		}
+	}
+}
+
+func (p *processor) consume(trace pdata.Traces) error {
+	rSpansSlice := trace.ResourceSpans()
+	for i := 0; i < rSpansSlice.Len(); i++ {
+		rSpan := rSpansSlice.At(i)
+
+		svc, ok := rSpan.Resource().Attributes().Get(conventions.AttributeServiceName)
+		if !ok {
+			return ErrNoServiceName
+		}
+
+		ilsSlice := rSpan.InstrumentationLibrarySpans()
+		for j := 0; j < ilsSlice.Len(); j++ {
+			ils := ilsSlice.At(j)
+
+			for k := 0; k < ils.Spans().Len(); k++ {
+				span := ils.Spans().At(k)
+
+				if p.store.ItemCount() >= p.maxItems {
+					return ErrTooManyItems
+				}
```

Comment:
That's a good point. I specially like the idea of trying to free the store if it reaches or it's close to reaching the limit.

> if we keep hitting maxItems, reduce the default expiration

That would affect the _correctness_ of the graphs, it could negatively affect the graphs more than not processing spans. It's something we can test with enough traffic.